mm2src/lp_ordermatch.rs

```diff
@@ -3285,6 +3341,130 @@ pub async fn set_price(ctx: MmArc, req: Json) -> Result<Response<Vec<u8>>, Strin
+    };
+
+    let min_base_amount = base_coin.min_trading_vol();
+    let min_rel_amount = rel_coin.min_trading_vol();
+
+    // Validate and Calculate Minimum Volume if min_volume is not in the request
+    let actual_min_base_vol = try_s!(validate_and_get_min_vol(
+        min_base_amount.clone(),
+        min_rel_amount.clone(),
+        req.min_volume.clone(),
+        req.new_price.clone().unwrap_or_else(|| my_order.clone().price)
+    ));
+
+    // Calculate order volume and add to update_msg if new_volume is found in the request
+    let volume = if req.max.unwrap_or(false) {
+        let max_volume = try_s!(get_max_volume(&ctx, &base_coin, &rel_coin).await);
+        update_msg = update_msg.with_new_max_volume(max_volume.clone().into());
+        max_volume
+    } else if Option::is_some(&req.new_volume) {
+        try_s!(
+            check_balance_for_maker_swap(
+                &ctx,
+                &base_coin,
+                &rel_coin,
+                req.new_volume.clone().unwrap(),
+                None,
+                None,
+                FeeApproxStage::OrderIssue
+            )
+            .await
+        );
+        update_msg = update_msg.with_new_max_volume(req.new_volume.clone().unwrap().into());
+        req.new_volume.unwrap()
+    } else {
+        my_order.max_base_vol.clone()
+    };
+
+    // Validate Order Volume
+    try_s!(validate_max_vol(
+        min_base_amount.clone(),
+        min_rel_amount.clone(),
+        volume.clone(),
+        req.min_volume.clone(),
+        req.new_price.clone().unwrap_or_else(|| my_order.clone().price)
+    ));
+
+    let conf_settings = OrderConfirmationsSettings {
+        base_confs: req
+            .base_confs
+            .unwrap_or_else(|| my_order.conf_settings.unwrap().base_confs),
+        base_nota: req
+            .base_nota
+            .unwrap_or_else(|| my_order.conf_settings.unwrap().base_nota),
+        rel_confs: req
+            .rel_confs
+            .unwrap_or_else(|| my_order.conf_settings.unwrap().rel_confs),
+        rel_nota: req.rel_nota.unwrap_or_else(|| my_order.conf_settings.unwrap().rel_nota),
+    };
+
+    update_msg = update_msg.with_new_conf_settings(conf_settings);
+
+    let updated_order = MakerOrder {
+        base: base.to_string(),
+        rel: rel.to_string(),
+        created_at: my_order.created_at,
+        updated_at: now_ms(),
+        max_base_vol: volume,
+        min_base_vol: actual_min_base_vol,
+        price: req.new_price.unwrap_or_else(|| my_order.price.clone()),
+        matches: my_order.matches,
+        started_swaps: my_order.started_swaps,
+        uuid: my_order.uuid,
+        conf_settings: Some(conf_settings),
+    };
+    save_my_maker_order(&ctx, &updated_order);
+
+    let rpc_result = MakerOrderForRpc::from(&updated_order);
+    let res = try_s!(json::to_vec(&json!({ "result": rpc_result })));
+
+    let mut my_maker_orders = ordermatch_ctx.my_maker_orders.lock().await;
+    my_maker_orders
```

Comment:
We can have a hypothetical but still possible situation here that the order is deleted at this point. It might be better to return the error in this case avoiding P2P message broadcast. Also, using `get_mut` and mutating only fields present in the request will allow avoiding resetting the fields as it happened with `min_base_vol`.

Reply:
I didn't use `get_mut` since I have the original order from here https://github.com/KomodoPlatform/atomicDEX-API/blob/ceb071130ad9ead90b4cf5239705c04d211de424/mm2src/lp_ordermatch.rs#L3349 that I used to get the original values for fields not present in the request. If there is more benefit for `get_mut` please let me know so I can use it.

Reply:
I think I incorrectly expressed my thought at the beginning, the main concern is not about `get_mut`:

1. As you clone the order initially you keep its old `matches` and `started_swaps` (let's call it `state A`).
2. During balance checks the order can be matched so its `matches` and `started_swaps` will be updated - it turns into `state B`.
3. After balance checks the current code resets the order to `state A` which is invalid. The information about the newly started swap will be lost. It turns into order available trading volume to be larger than initially expected by the user, which can be considered as financial loss in some circumstances. E.g. if I prefer to sell only 1 BTC but the software actually sells more I won't be happy :slightly_smiling_face: 

The mitigation proposal is as follows:
Replace 
```rust
let my_order = match my_maker_orders.get(&req.uuid) {
        Some(order) => order.clone(),
        None => return ERR!("There is no order with UUID {}", &req.uuid),
    };

    drop(my_maker_orders);
```
With 
```rust
    let (base_coin, rel_coin) = match my_maker_orders.get(&req.uuid) {
        Some(order) => {
            let base = order.base.as_str();
            let base_coin: MmCoinEnum = match try_s!(lp_coinfind(&ctx, base).await) {
                Some(coin) => coin,
                None => return ERR!("Base coin {} has been removed from config", base),
            };

            let rel = order.rel.as_str();
            let rel_coin: MmCoinEnum = match try_s!(lp_coinfind(&ctx, rel).await) {
                Some(coin) => coin,
                None => return ERR!("Rel coin {} has been removed from config", rel),
            };
        },
        None => return ERR!("There is no order with UUID {}", &req.uuid),
    };
```

Further, refactor the 
```rust
    let _original_order = match my_maker_orders.entry(updated_order.uuid) {
        Entry::Vacant(_) => return ERR!("Order with UUID: {} has been deleted", updated_order.uuid),
        Entry::Occupied(mut entry) => entry.insert(updated_order),
    };
```
To something like
```rust
    match my_maker_orders.get_mut(&req.uuid) {
        None => return ERR!("Order with UUID: {} has been deleted", req.uuid),
        Some(order) => order.apply_update(...),
    };
```

Where `MakerOrder::apply_update(&mut self, ...)` takes either multiple args or a struct with optional fields to update. Please ensure that it updates only fields present in the request. Order recreation leads to an invalid state.